stream_alert/classifier/payload/s3.py

```diff
@@ -0,0 +1,214 @@
+        # Use the urllib unquote method to decode any url encoded characters
+        # (ie - %26 --> &) from the bucket and key names
+        return urllib.unquote(data).decode('utf-8')
+
+    def _check_size(self):
+        """Ensure the S3 file's size is not too large to download into the Lambda environment
+
+        Returns:
+            bool: True if the file is smaller than 128 MB, False otherwise
+        """
+        # size == 0 or greater than 128MB
+        if self.size == 0 or (self.size > 128 * 1024 * 1024):
+            raise S3PayloadError('S3 object {}/{} has an invalid size and cannot be downloaded'
+                                 'from S3: {}'.format(self.bucket, self.key, self.display_size))
+
+    @staticmethod
+    def _shred_temp_directory():
+        """Delete all objects in the container's temp directory"""
+        LOGGER.debug('Shredding temp directory')
+
+        for root, dirs, files in os.walk(tempfile.gettempdir(), topdown=False):
+            for name in files:
+                subprocess.check_call([  # nosec
+                    'shred', '--force', '--iterations=1',
+                    '--remove', os.path.join(root, name)])
+            for name in dirs:
+                os.rmdir(os.path.join(root, name))  # nosec
+
+    @staticmethod
+    def _gz_reader(open_file):
+        open_file.seek(0)
+        reader = gzip.GzipFile(fileobj=open_file, mode='r')
+        try:
+            # Test to ensure this is gzip data, then rewind
+            reader.read(1)
+            reader.rewind()
+        except IOError:
+            # Fall back on the default reader
+            reader = open_file
+            reader.seek(0)
+
+        # Return either the gzip reader or the original reader
+        return reader
+
+    @staticmethod
+    def _jsonlines_reader(open_file):
+        open_file.seek(0)
+        reader = None
+        try:
+            json_lines = jsonlines.Reader(open_file)
+            json_lines.read()
+            reader = json_lines
+        except ValueError:
+            reader = open_file
+
+        open_file.seek(0)
+        return reader
+
+    @classmethod
+    def _read_downloaded_object(cls, open_file):
+        """Read the contents of the downloaded S3 file using the open file handle
+
+        Args:
+            open_file (file-like object): File handle from which to read the data
+
+        Yields:
+            tuple: line number, contents of the line being read
+        """
+        reader = cls._gz_reader(open_file)
+        try:
+            # Try to just load the reader as regular json first
+            yield 1, json.load(reader)
+            return
+        except ValueError:
+            pass
+
+        # Iterate over the lines, returning each
+        # This could be dicts from a jsonlines.Reader, or raw strings
+        for line_num, line in enumerate(cls._jsonlines_reader(reader), start=1):
+            yield line_num, line.strip() if isinstance(line, basestring) else line
```

Comment:
It will yield raw string instead of a tuple (line number, content) if the `line` is not a `basestring`. Is it expected?

Reply:
discussed offline